### ISSUE TITLE
Image intensity scaling on Philips systems

### ISSUE THREAD

**Issue by dchristiaens** [MRtrix3/mrtrix3#2499]
Consecutive Philips dMRI scans often have inconsistent intensity scaling. We can often fix this in post-processing with `dwicat`, but not, for example, for data acquired with different TE. I would therefore like to retrieve the exact scaling from the headers.

I had a look at the Dicom tags of some of our Philips data to see if we could reverse-engineer what they do. In 3ee008050accff30cb6bc8f16484e6cf695d5791, I propose a change that appears to fix the scaling for 3 datasets I could test with (same scanner).

Besides the `RescaleSlope` used by default, the dMRI Dicoms also have a tag `MRScaleSlope` in the diffusion group (0x2005). For instance:
```
[DCM] 0028 1052 DS        2    13332       RescaleIntercept                   0
[DCM] 0028 1053 DS       16    13342       RescaleSlope                       1.08083028083028
[DCM] 0028 1054 LO       10    13366       RescaleType                        normalized
...
[DCM] 2005 100D FL        4    13676       MRScaleIntercept                   0
[DCM] 2005 100E FL        4    13688       MRScaleSlope                       0.0190212
```
In the data I tested, it seems that the `RescaleSlope` is to be ignored. Instead, the correct scaling seems to be _the inverse of_ `MRScaleSlope`.

A big disclaimer is that this was tested only for 1 scanner, 2 software versions, 3 datasets. So I am not sufficiently confident about this result to merge it into master anytime soon. I simply thought I'd document what I found here, if someone wants to test it further.

**Comment by Lestropie** [comment]
Eeek. I've no idea how common or reliable that field is; probably need to wait for @jdtournier to go through the private DICOM test data.

Hypothetically, if it were to be the case that it's not possible to have a single heuristic to determine scaling across DICOMs, then `dwicat` may still be the right place for the functionality; just that it would take two drastically different code branches: one being intensity rescaling based on histogram matching, one being intensity rescaling based on the contents of `0x2005`.

**Comment by neurolabusc** [comment]
Relevant to this is [Philips image scaling notes](https://github.com/rordenlab/dcm2niix/tree/master/Philips#image-scaling). The first draft was written by a Philips Senior Clinical Scientist (@drmclem). It is worth noting that some Philips
 DICOMs use different scaling values for different 2D slices in a volume, where NIfTI requires a single slope and intercept for all voxels in the series. This explains why dcm2niix can be forced to up-sample some Philips data (e.g. from 16-bit integers to 32-bit floats). You can find small [validation datasets here](https://github.com/neurolabusc/dcm_qa_philips) and larger DWI-specific datasets [here](https://www.nitrc.org/plugins/mwiki/index.php/dcm2nii:MainPage#Diffusion_Tensor_Imaging).

**Comment by jdtournier** [comment]
> It is worth noting that some Philips DICOMs use different scaling values for different 2D slices in a volume, where NIfTI requires a single slope and intercept for all voxels in the series.

Thanks Chris. We added support for this back in 2018 with a4134de7, that bit should be OK. 

Otherwise, I remember trying to do this a while back, based on a Philips document I'd been sent at the time. The relevant slide in that document states:
```
FP = ( SV / SS ) + ( RI / (RS × SS) )
```

where:

- `FP`: “Floating-point Value”; reflects the MR signal range directly after reconstruction. This format should be used when quantitatively comparing multiple acquisitions.
- `SV`: “Stored Value”; stored value in DICOM image file.
- `SS`: “MR Scale Slope”; from DICOM header tag: (2005, 100E).
- `RI`: “Rescale Intercept”; from DICOM header tag: (0028, 1052)
- `RS`: “Rescale Slope”; from DICOM header tag: (0028, 1053)

However, in my experience, that didn't really work - there were still residual differences in scaling across acquisitions, in some cases larger than the original differences (without these changes applied). Happy to revisit though... 

> probably need to wait for @jdtournier to go through the private DICOM test data.

Ouch. That's likely to invalidate all (or at least much of) the Philips test data we currently have... Not sure I'm looking forward to updating all of that.

**Comment by dchristiaens** [comment]
Thanks @neurolabusc for your input - the `dcm2niix` docs on this topic are very helpful. Our local Philips clinical scientist shared a document with me (presumably the same one Donald has) that describes the same intensity scaling. 

@treanus has extensively tested the recent `dcm2niix` version in recent and legacy data acquired locally on several different Philips scanners and software versions. He found the image intensity to be much more stable and consistent across scans, so I reckon it's safe to assume that the `dcm2niix` implementation is correct. @jdtournier, as you seem to have experienced differently, any chance you could give it another look? 

I have implemented the modified scaling in bfa7c82 and suggest we merge this into `dev` if everyone's happy.

**Comment by jdtournier** [comment]
Happy in principle, though I have a couple of practical concerns:

- Have you verified that these changes replicate the output from `dcm2niix` for a realistic selection of datasets? 

- The current changes depend on the order of presentation of the relevant DICOM elements. This only works if the standard DICOM tags appear first in the file, so the Philips tags can be used to update those values later. I expect this will probably be the case in 99% of cases, but it's nonetheless a bit fragile as a solution... To do this properly, I think we might need to store the Philips tags separately, and update the scaling parameters after parsing is done... 

Otherwise, as I mentioned before, this will probably break all my existing Philips test data, and it's going to be painful to update all of that to match... But then, I have other tests to add to the suite, so I'll just need to bite the bullet when that happens.

**Comment by dchristiaens** [comment]
> * Have you verified that these changes replicate the output from `dcm2niix` for a realistic selection of datasets?

Yes, I have for a few recent dMRI and T1w datasets (all from the same scanner and software version). @treanus will verify on a larger set of legacy data.

> * The current changes depend on the order of presentation of the relevant DICOM elements. This only works if the standard DICOM tags appear first in the file, so the Philips tags can be used to update those values later. I expect this will probably be the case in 99% of cases, but it's nonetheless a bit fragile as a solution... To do this properly, I think we might need to store the Philips tags separately, and update the scaling parameters after parsing is done...

This sounds like a very important point, that clearly reveals my limited knowledge of the DICOM format. I'll have a look and get back to you if I can't figure it out.

**Comment by neurolabusc** [comment]
@dchristiaens since @jdtournier is unable to share his validation dataset, can I suggest that a user with access to a Philips scanner can acquire public validation datasets with validated solutions? I provide several dcm_qa* validation datasets, including [dcm_qa_philips](https://github.com/neurolabusc/dcm_qa_philips) that allow me to regression test dcm2niix. Other examples of datasets are [here](https://www.nitrc.org/plugins/mwiki/index.php/dcm2nii:MainPage#Archival_MRI). However, since I do not have access to Philips hardware, I can not acquire calibrated images that showcase the [Philips image scaling](https://github.com/rordenlab/dcm2niix/tree/master/Philips#image-scaling). My IRB does not consider development images of scientists as research, so I can share them. However, for this issue samples of a phantom would suffice (and where possible it helps to use low resolution, small images as cross-platform continuous integration mechanisms like AppVeyor are slow).
